### PR TITLE
Add additional prov tests

### PR DESCRIPTION
**Pull request by hihigupt**

### CODE REVIEW

File: tests/azure_iot_provisioning_client_ut.c [hunk 1]
```diff
@@ -583,8 +605,14 @@ static void testAzureIoTProvisioningClient_Register_RegistrationFailure( void **
...
-    xResult = AzureIoTProvisioningClient_Register( &xTestProvisioningClient, 0 );
-    assert_int_equal( xResult, eAzureIoTProvisioningServerError );
+    assert_int_equal( AzureIoTProvisioningClient_Register( &xTestProvisioningClient, 0 ),
+                      eAzureIoTProvisioningServerError );
+
+    /* Read extended code */
+    assert_int_equal( AzureIoTProvisioningClient_GetExtendedCode( &xTestProvisioningClient,
+                                                                  &ulCode ),
+                      eAzureIoTProvisioningSuccess );
+    assert_int_equal( ulExtendedCode, ulCode );
```
Comment (danewalton) [inline tests/azure_iot_provisioning_client_ut.c:615]: should this be something like
```suggestion
    assert_int_equal( ulExpectedExtendedCode, ulExtendedCode );
```

File: tests/azure_iot_provisioning_client_ut.c [hunk 2]
```diff
@@ -637,6 +665,64 @@ static void testAzureIoTProvisioningClient_Register_QueryFailure( void ** ppvSta
...
+
+    /* AckSubscribe */
+    prvRegistrationAckSubscribeStep( &xTestProvisioningClient );
+
+    /* Publish Registration Request */
+    prvRegistrationPublishStep( &xTestProvisioningClient );
+
+    /* Registration response */
+    prvGenerateGoodResponse( &xPublishInfo, 1 );
+    assert_int_equal( AzureIoTProvisioningClient_Register( &xTestProvisioningClient, 0 ),
```
Comment (danewalton) [inline tests/azure_iot_provisioning_client_ut.c:691]: nit: can we have a macro for `0` in the test file for no wait? ex
```suggestion
    assert_int_equal( AzureIoTProvisioningClient_Register( &xTestProvisioningClient, provtestNO_WAIT_TIMEOUT ),
```
easier to read and remember what that param is doing

**Review by danewalton — APPROVED** [review]

File: tests/azure_iot_provisioning_client_ut.c (hunk 2 above)
Comment (hihigupt) [reply tests/azure_iot_provisioning_client_ut.c:691]: I think better will be to have marco from the provisioning header.

**Review by jspaith — APPROVED** [review]

File: tests/azure_iot_provisioning_client_ut.c [hunk 3]
```diff
@@ -53,6 +54,13 @@ static const uint8_t ucFailureHubResponse[] = "{ \
           \"etag\":\"XXXXXXXXXXX=\"\
          }\
 }";
+static const uint8_t ucInvalidHubResponse[] = "{ \
```
[the remark this answers is not in the document]
Comment (hihigupt) [reply tests/azure_iot_provisioning_client_ut.c:57]: Sure,

---
In reply to: [608800106](https://github.com/Azure/azure-iot-middleware-freertos/pull/68#discussion_r608800106) [](ancestors = 608800106)

File: tests/azure_iot_provisioning_client_ut.c [hunk 4]
```diff
@@ -578,25 +617,93 @@ static void testAzureIoTProvisioningClient_Register_QueryFailure( void ** ppvSta
...
+    xPacketInfo.ucType = 0;
+    assert_int_equal( AzureIoTProvisioningClient_Register( &xTestProvisioningClient,
+                                                           azureiotPROVISIONING_NO_WAIT ),
+                      eAzureIoTProvisioningPending );
 
+    /* Invalid registration response */
+    prvGenerateResponse( &xPublishInfo, 0,
+                         ucInvalidHubResponse,
+                         sizeof( ucInvalidHubResponse ) - 1 );
+    assert_int_equal( AzureIoTProvisioningClient_Register( &xTestProvisioningClient,
```
[the remark this answers is not in the document]
Comment (hihigupt) [reply tests/azure_iot_provisioning_client_ut.c:699]: 698: seed MQTT payload 
699: push payload to middleware layer via callback inside process_loop of MQTT.
704: parse payload given to this layer.

Reason 699 do not do parsing is that is in context of callback, and I do not want to hijack that callback.

---
In reply to: [608803910](https://github.com/Azure/azure-iot-middleware-freertos/pull/68#discussion_r608803910) [](ancestors = 608803910)

File: tests/azure_iot_provisioning_client_ut.c [hunk 5]
```diff
@@ -673,6 +780,47 @@ static void testAzureIoTProvisioningClient_GetDeviceAndHub_Success( void ** ppvS
 }
 /*-----------------------------------------------------------*/
 
+static void testAzureIoTProvisioningClient_WithCustomPayload_Failure( void ** ppvState )
```
[the remark this answers is not in the document]
Comment (hihigupt) [reply tests/azure_iot_provisioning_client_ut.c:783]: line 810 is failure caused in registration. 

---
In reply to: [608805798](https://github.com/Azure/azure-iot-middleware-freertos/pull/68#discussion_r608805798) [](ancestors = 608805798)

File: tests/azure_iot_provisioning_client_ut.c [hunk 6]
```diff
@@ -673,6 +780,47 @@ static void testAzureIoTProvisioningClient_GetDeviceAndHub_Success( void ** ppvS
...
+    /* AckSubscribe */
+    prvRegistrationAckSubscribeStep( &xTestProvisioningClient );
+
+    /* Publish payload generation failed */
+    assert_int_equal( AzureIoTProvisioningClient_Register( &xTestProvisioningClient,
+                                                           azureiotPROVISIONING_NO_WAIT ),
+                      eAzureIoTProvisioningFailed );
+
+    /* After failed registration, AzureIoTProvisioningClient_GetDeviceAndHub should return error */
+    assert_int_not_equal( AzureIoTProvisioningClient_GetDeviceAndHub( &xTestProvisioningClient,
```
[the remark this answers is not in the document]
Comment (hihigupt) [reply tests/azure_iot_provisioning_client_ut.c:814]: I do not think we should zero out value if in error case. Caller should not take any dependency on the out parameter if API fails.

**Pull request merged**